kernel/src/utilities/streaming_process_slice.rs

```diff
@@ -0,0 +1,269 @@
+/// 2. The first buffer is prepared according to the format below. The `flags`
+///    field's version bits are set to `0`. The process clears the `exceeded`
+///    flag. It may set or clear the `halt` flag. All reserved flags must be set
+///    to `0`. Finally, the `offset` bytes (intepreted as a u32 value in native
+///    endianness) are set to `0`.
+///
+/// 3. The process `allow`s this buffer to a kernel driver.
+///
+/// 4. The kernel driver writes incoming data starting at the `data` field +
+///    `offset` bytes. After each write, the kernel increments `offset` by the
+///    number of bytes written.
+///
+///    For each *chunk* written to the buffer (where a *chunk* is an
+///    application-defined construct, such as a network packet), the kernel only
+///    increments `offset` if the full chunk was successfully written into the
+///    buffer. The kernel may or may not modify any data after the current
+///    `offset` value, regardless of whether any metadata was updated. The
+///    kernel never modifies any data in the region of
+///    `[data.start; data.start + offset)`.
+///
+///    Should the write of a chunk fail because the buffer has insufficient
+///    space left, the kernel will set the `exceeded` flag bit (index 0).
+///
+///    The `halt` flag bit as set by the process governs the kernel's behavior
+///    once the `exceeded` flag is set: if `halt` is cleared, the kernel will
+///    attempt to write future, smaller chunks to be buffer (and thus implicitly
+///    discarding some packets). If `halt` and `exceeded` are both set, the
+///    kernel will stop writing any data into the buffer.
+///
+/// 5. The kernel will schedule an upcall to the process, indicating that a
+///    write to the buffer (or setting the `exceeded`) flag occurred. The kernel
+///    may schedule only one upcall for the first chunk written to the buffer,
+///    or multiple upcalls (e.g., one upcall per chunk written). A process must
+///    not rely on the number of upcalls received and instead rely on the buffer
+///    metadata (`offset` and the `flags` bits) to determine the amount of data
+///    written to the buffer.
+///
+/// 6. The process prepares its second buffer, following step 2. The process
+///    then issues an `allow` operation that atomically swaps the current
+///    allowed buffer by its second buffer.
+///
+/// 7. The process can now process the received chunks contained in the initial
+///    buffer, while the kernel receives new chunks in the other, newly allowed
+///    buffer.
+///
+///
+/// The version 0 buffer format is specified as follows:
+/// ```text,ignore
+/// 0          1          2          3          4          5
+/// +----------+----------+----------+----------+----------+-------------------...
+/// | flags    | buffer offset                             | data
+/// +----------+----------+----------+----------+----------+-------------------...
+/// | VVVVxxHE | 32 bits (native endian)                   |
```

Comment:
It feels like we would be better off with the offset as 3 bytes to help with align `data` to a multiple of 4. It's possible it won't matter, and userspace might align the entire allow buffer arbitrarily, and userspace may only use data as a `u8`, but it might avoid a potential issue to have data start after 4 bytes.

Reply:
> It's possible it won't matter, and userspace might align the entire allow buffer arbitrarily, and userspace may only use data as a `u8`,

Right. The kernel can't rely on any alignment, and userspace can simply add three addl. bytes of padding if it wants data to be aligned. We need to specify a fixed width for this field, and 32 bits just seems less arbitrary than 24.

Reply:
How can userspace add three additional bytes of padding? Before this entire buffer? In that case I agree, but I don't see why we would set it up that way unless we expect 17 MB+ buffers.

Reply:
Exactly, it could simply shift this entire buffer by three bytes within its allocation, in which case both offset and data were aligned to 4 bytes.

I don't think that 17MB is an exorbitant size for these buffers on larger platforms (hello, aarch64 and x86 support!), and having a proper integer type that can communicate the limitation of 32-bit offset values (i.e. `u32` / `uint32_t`) has me strongly in favor of that vs 24-bit and runtime errors deep within this abstraction.

Reply:
This has a small downside, we limit the buffer to 16 MB. While on small devices this is fine, on larger devices such as the x86 this might be too restrictive.

Reply:
What if we make version 4 bytes then? If we are designing around multi-megabyte buffers, then 3 bytes shouldn't matter.

Reply:
I'm fine with that.